pkg/resource/deploy/step_generator.go

```diff
@@ -1131,6 +1154,19 @@ func (sg *stepGenerator) loadResourceProvider(
 	return p, nil
 }
 
+func (sg *stepGenerator) getProviderResource(urn resource.URN, provider string) (*resource.State, result.Result) {
+	if provider == "" {
+		return nil, nil
+	}
+
+	ref, err := providers.ParseReference(provider)
+	if err != nil {
+		sg.plan.Diag().Errorf(diag.GetBadProviderError(urn), provider, urn, err)
+		return nil, result.Bail()
+	}
+	return sg.providers[ref.URN()], nil
```

Comment:
Just for belt-and-suspenders type check, can we wire in a new error if for whatever reason `ref.URN()` isn't found within the `sg.providers` slice? Given the number of moving parts, I wouldn't be surprised if there is some weird unexpected corner case that would lead to that being the case.

Reply:
We may simply want to assert that
a) there is no error when parsing the provider reference, and
b) the provider exists in the map

I believe that all callers of this method are on paths that have previously validated those properties.

Reply:
> I believe that all callers of this method are on paths that have previously validated those properties.

Yep. Added asserts.